Lab | Revisiting Machine Learning Case Study

Complete the following steps on the categorical columns in the dataset:
1.Check for null values in all the columns
2.Exclude the following variables by looking at the definitions. Create a new empty list called drop_list. We will append this list and then drop all the columns in this list later:
OSOURCE - symbol definitions not provided, too many categories
ZIP - we are including state already
Remove those columns from the dataframe
Perform all of the cleaning processes from the Lesson.
Reduce the number of categories in the column GENDER. The column should only have either "M" for males, "F" for females, and "other" for all the rest

Note that there are a few null values in the column. We will first replace those null values using the code below:

print(categorical['GENDER'].value_counts())
categorical['GENDER'] = categorical['GENDER'].fillna('F')

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 
pd.set_option('display.max_columns', None)


In [2]:
data = pd.read_csv('learningSet.csv')

C:\Users\torra\AppData\Local\Temp\ipykernel_8776\3673836359.py:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('learningSet.csv')


In [3]:
#Check for null values in all the columns
#First we check if there are any null values in our dataset
data.isnull().values.any()
#Now we know that we have a problem.

True

In [4]:
#Next we are going to check in which columns these null values are located.
#We are going to express it in a percentage.
#We want to know how many null values are compare to the total inputs in the column
data.isna().sum()/len(data)

ODATEDW     0.000000
OSOURCE     0.000000
TCODE       0.000000
STATE       0.000000
ZIP         0.000000
              ...   
MDMAUD_R    0.000000
MDMAUD_F    0.000000
MDMAUD_A    0.000000
CLUSTER2    0.001383
GEOCODE2    0.001383
Length: 481, dtype: float64

In [5]:
#we are going to create a new dataframe containing just the column name and the % of null values.
nulls_percent_df = pd.DataFrame(data.isna().sum()/len(data)).reset_index()
nulls_percent_df
nulls_percent_df.columns = ['column_name', 'nulls_percentage']
nulls_percent_df

column_name  nulls_percentage
0       ODATEDW          0.000000
1       OSOURCE          0.000000
2         TCODE          0.000000
3         STATE          0.000000
4           ZIP          0.000000
..          ...               ...
476    MDMAUD_R          0.000000
477    MDMAUD_F          0.000000
478    MDMAUD_A          0.000000
479    CLUSTER2          0.001383
480    GEOCODE2          0.001383

[481 rows x 2 columns]

In [6]:
#We want to analyze those columns who have null values
nulls_percent_df[nulls_percent_df['nulls_percentage']!=0]

column_name  nulls_percentage
16          AGE          0.248030
23      NUMCHLD          0.870184
24       INCOME          0.223096
26      WEALTH1          0.468830
28      MBCRAFT          0.553955
..          ...               ...
455    RAMNT_24          0.814090
466    NEXTDATE          0.104526
467     TIMELAG          0.104526
479    CLUSTER2          0.001383
480    GEOCODE2          0.001383

[92 rows x 2 columns]

originally we analyzed all the columns of the data set (481), of those 92 present null values.
The number of columns presenting null values is considerable (19.12%)

In [7]:
#We are going to check the head and tail of the data frame
nulls_percent_df.head(25)
print(nulls_percent_df.head(25))

   column_name  nulls_percentage
0      ODATEDW          0.000000
1      OSOURCE          0.000000
2        TCODE          0.000000
3        STATE          0.000000
4          ZIP          0.000000
5     MAILCODE          0.000000
6     PVASTATE          0.000000
7          DOB          0.000000
8       NOEXCH          0.000000
9     RECINHSE          0.000000
10       RECP3          0.000000
11     RECPGVG          0.000000
12    RECSWEEP          0.000000
13      MDMAUD          0.000000
14      DOMAIN          0.000000
15     CLUSTER          0.000000
16         AGE          0.248030
17     AGEFLAG          0.000000
18    HOMEOWNR          0.000000
19     CHILD03          0.000000
20     CHILD07          0.000000
21     CHILD12          0.000000
22     CHILD18          0.000000
23     NUMCHLD          0.870184
24      INCOME          0.223096


In [8]:
nulls_percent_df.tail(25)
print(nulls_percent_df.tail(25))

    column_name  nulls_percentage
456    RAMNTALL          0.000000
457    NGIFTALL          0.000000
458    CARDGIFT          0.000000
459    MINRAMNT          0.000000
460    MINRDATE          0.000000
461    MAXRAMNT          0.000000
462    MAXRDATE          0.000000
463    LASTGIFT          0.000000
464    LASTDATE          0.000000
465    FISTDATE          0.000000
466    NEXTDATE          0.104526
467     TIMELAG          0.104526
468     AVGGIFT          0.000000
469    CONTROLN          0.000000
470    TARGET_B          0.000000
471    TARGET_D          0.000000
472    HPHONE_D          0.000000
473      RFA_2R          0.000000
474      RFA_2F          0.000000
475      RFA_2A          0.000000
476    MDMAUD_R          0.000000
477    MDMAUD_F          0.000000
478    MDMAUD_A          0.000000
479    CLUSTER2          0.001383
480    GEOCODE2          0.001383


We are going to consider a threshold of 25% percent of null values, if the column has a larger value, it's a candidate for elimination.

In [9]:
columns_above_threshold = nulls_percent_df[nulls_percent_df['nulls_percentage']>0.25]
columns_above_threshold['column_name']
print(columns_above_threshold)

    column_name  nulls_percentage
23      NUMCHLD          0.870184
26      WEALTH1          0.468830
28      MBCRAFT          0.553955
29     MBGARDEN          0.553955
30      MBBOOKS          0.553955
..          ...               ...
451    RAMNT_20          0.917327
452    RAMNT_21          0.900296
453    RAMNT_22          0.781233
454    RAMNT_23          0.917631
455    RAMNT_24          0.814090

[72 rows x 2 columns]


Of the initial 92 columns with null values, we found 72 with null values over the treshold.
In percentage terms, it means that 78.26% of the columns with null values are over the threshold.
As it is a large number of columns, we are going to look in to each major group and see if any is worth saving.


The following fields are from the giving history file (RDATE and RAMNT), the first is related to the dates the donation were made and the second to the dollar amount donated in each mailing campaign.
We think that we don't need this level of capilarity, and we have more interesting variables related to this topic such as RAMNTALL and LASTDATE.
The fields starting with ADATE correspond to the specifics dates the mailing campaign took place, this information is irrelevant for our analysis.
The following columns (MBCRAFT', 'MBGARDEN', 'MBBOOKS', 'MBCOLECT', 'MAGFAML', 'MAGFEM', 'MAGMALE', 'PUBGARDN', 'PUBCULIN', 'PUBHLTH', 'PUBDOITY', 'PUBNEWFN', 'PUBPHOTO', 'PUBOPP') indicate the number known times the donor has responded to othertypes of mail order offers. Eventhought it may seem as interesting information to have it's to granular. We allready have a variable as HIT that Indicates total number of known times the donor has
responded to a mail order offer other than PVA's.
'WEALTH1' and WEALTH2 are wealth ratings, of the second we have a detail description.
WEALTH2:Wealth rating uses median family income and population statistics from each area to index relative wealth within each state
The segments are denoted 0-9, with 9 beingthe highest income group and zero being the lowest.
We are going to keep both wealth indexes, and analyze a little further if we can drop one of them.
We are going to drop 'NUMCHLD'(number of children), we have other variables that measure the children in the household.


In [10]:
#We create a list with the columns to drop (those above the threshold)
drop_columns_list= list(columns_above_threshold['column_name'])
print(drop_columns_list)

['NUMCHLD', 'WEALTH1', 'MBCRAFT', 'MBGARDEN', 'MBBOOKS', 'MBCOLECT', 'MAGFAML', 'MAGFEM', 'MAGMALE', 'PUBGARDN', 'PUBCULIN', 'PUBHLTH', 'PUBDOITY', 'PUBNEWFN', 'PUBPHOTO', 'PUBOPP', 'WEALTH2', 'ADATE_5', 'ADATE_10', 'ADATE_13', 'ADATE_15', 'ADATE_17', 'ADATE_19', 'ADATE_20', 'ADATE_21', 'ADATE_22', 'ADATE_23', 'ADATE_24', 'RDATE_3', 'RDATE_4', 'RDATE_5', 'RDATE_6', 'RDATE_7', 'RDATE_8', 'RDATE_9', 'RDATE_10', 'RDATE_11', 'RDATE_12', 'RDATE_13', 'RDATE_14', 'RDATE_15', 'RDATE_16', 'RDATE_17', 'RDATE_18', 'RDATE_19', 'RDATE_20', 'RDATE_21', 'RDATE_22', 'RDATE_23', 'RDATE_24', 'RAMNT_3', 'RAMNT_4', 'RAMNT_5', 'RAMNT_6', 'RAMNT_7', 'RAMNT_8', 'RAMNT_9', 'RAMNT_10', 'RAMNT_11', 'RAMNT_12', 'RAMNT_13', 'RAMNT_14', 'RAMNT_15', 'RAMNT_16', 'RAMNT_17', 'RAMNT_18', 'RAMNT_19', 'RAMNT_20', 'RAMNT_21', 'RAMNT_22', 'RAMNT_23', 'RAMNT_24']


In [11]:

drop_columns_list.remove('WEALTH1')
drop_columns_list.remove('WEALTH2')


We need to include in the drop list the columns: OSOURCE and ZIP 


In [12]:
drop_columns_list.append('OSOURCE')
drop_columns_list.append('ZIP')

In [13]:
print(drop_columns_list)

['NUMCHLD', 'MBCRAFT', 'MBGARDEN', 'MBBOOKS', 'MBCOLECT', 'MAGFAML', 'MAGFEM', 'MAGMALE', 'PUBGARDN', 'PUBCULIN', 'PUBHLTH', 'PUBDOITY', 'PUBNEWFN', 'PUBPHOTO', 'PUBOPP', 'ADATE_5', 'ADATE_10', 'ADATE_13', 'ADATE_15', 'ADATE_17', 'ADATE_19', 'ADATE_20', 'ADATE_21', 'ADATE_22', 'ADATE_23', 'ADATE_24', 'RDATE_3', 'RDATE_4', 'RDATE_5', 'RDATE_6', 'RDATE_7', 'RDATE_8', 'RDATE_9', 'RDATE_10', 'RDATE_11', 'RDATE_12', 'RDATE_13', 'RDATE_14', 'RDATE_15', 'RDATE_16', 'RDATE_17', 'RDATE_18', 'RDATE_19', 'RDATE_20', 'RDATE_21', 'RDATE_22', 'RDATE_23', 'RDATE_24', 'RAMNT_3', 'RAMNT_4', 'RAMNT_5', 'RAMNT_6', 'RAMNT_7', 'RAMNT_8', 'RAMNT_9', 'RAMNT_10', 'RAMNT_11', 'RAMNT_12', 'RAMNT_13', 'RAMNT_14', 'RAMNT_15', 'RAMNT_16', 'RAMNT_17', 'RAMNT_18', 'RAMNT_19', 'RAMNT_20', 'RAMNT_21', 'RAMNT_22', 'RAMNT_23', 'RAMNT_24', 'OSOURCE', 'ZIP']


In [23]:
drop_columns_df=pd.DataFrame(drop_columns_list)

In [30]:

data['GENDER'] = data['GENDER'].fillna('F')
print(data['GENDER'].value_counts())

GENDER
F    51277
M    39094
      2957
U     1715
J      365
C        2
A        2
Name: count, dtype: int64


In [31]:
data['GENDER']=data['GENDER'].replace(['J', 'C','A'], 'U', inplace=True)

Lab | Feature engineering

In this lab, you will use learningSet.csv file which you have already cloned in the previous activities.
Continue working in the same notebook as you did in the previous Lab.
Again go through all of the Numerical columns and apply the techniques that were performed in the lesson



Then we will work on cleaning some of the other columns in the dataset using the techniques that we used before in the lessons.
Check for null values in the numerical columns.
After going through the lesson techniques there should only be a few columns left with NaN values to clean.
Use appropriate methods to clean the columns which still contain NaN values.
Use appropriate EDA technique where ever necessary.